File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PropertiesAccessCorrectField.vb
Imports System
Imports GalaSoft.MvvmLight

Namespace Tests.Diagnostics
    Class NonCompliantClass_FromRspec
        Private x_ As Integer
        Private y_ As Integer

        Public Property X As Integer
            Get
                Return x_
            End Get
            Set(ByVal value As Integer)
                x_ = value
            End Set
        End Property

        Public Property Y As Integer
            Get
                Return x_ ' Noncompliant {{Refactor this getter so that it actually refers to the field 'y_'.}}
            End Get
            Set(ByVal value As Integer)
                x_ = value ' Noncompliant {{Refactor this setter so that it actually refers to the field 'y_'.}}
            End Set
        End Property
    End Class

    Structure NonCompliantStruct_FromRspec
        Private x_ As Integer
        Private y_ As Integer

        Public Property X As Integer
            Get
                Return x_
            End Get
            Set(ByVal value As Integer)
                x_ = value
            End Set
        End Property

        Public Property Y As Integer
            Get
                Return x_ ' Noncompliant
            End Get
            Set(ByVal value As Integer)
                x_ = value ' Noncompliant
            End Set
        End Property
    End Structure

    Class NonCompliant_UnderscoresInNamesAndCasing
        Private yyy As Integer
        Private __x__X As Integer

        Public Property XX As Integer ' test that underscores and casing in names are ignored
            Get
                Return yyy ' Noncompliant {{Refactor this getter so that it actually refers to the field '__x__X'.}}
            End Get
            Set(ByVal value As Integer)
                yyy = value ' Noncompliant {{Refactor this setter so that it actually refers to the field '__x__X'.}}
            End Set
        End Property

        Public ReadOnly Property _Y___Y_Y_ As String
            Get
                Return __x__X ' Noncompliant
            End Get
        End Property
    End Class

    Class NonCompliant_FieldTypeIsIgnored
        Private aaa_ As Integer
        Private aString As String

        Public Property AAA As String
            Get
                Return aString ' Noncompliant - field called 'aaa' exists, even though type is different
            End Get
            Set(ByVal value As String)
                aString = value ' Noncompliant
            End Set
        End Property
    End Class

    Class NonCompliant_AssigningToExpression
        Private aaa_ As Integer
        Private aString As String

        Public WriteOnly Property AAA As String
            Set(ByVal value As String)
                aString = "foo" & value ' Noncompliant
            End Set
        End Property
    End Class

    Partial Class NonCompliant_PartialClass
        Private myProperty_ As Object
    End Class

    Partial Class NonCompliant_PartialClass
        Private anotherObject As Object
    End Class

    Partial Class NonCompliant_PartialClass
        Public Property MyProperty As Object
            Get
                Return Me.anotherObject ' Noncompliant
            End Get
            Set(ByVal value As Object)
                Me.anotherObject = value ' Noncompliant
            End Set
        End Property
    End Class

    Class NonCompliant_ComplexProperty
        Private field1_ As Integer
        Private field2_ As Integer
        Private initialized As Boolean
        Private isDisposed As Boolean

        Public Property Field1 As Integer
            Get ' Noncompliant
'           ^^^
                If Not Me.initialized Then
                    Throw New InvalidOperationException()
                End If

                If Me.isDisposed Then
                    Throw New ObjectDisposedException("object name")
                End If

                Return Me.field2_
            End Get
            Set(ByVal value As Integer)

                If value < 0 Then
                    Throw New ArgumentOutOfRangeException(NameOf(value))
                End If

                Me.field2_ = value ' Noncompliant
'                  ^^^^^^^
            End Set
        End Property
    End Class

    Class NonCompliant_Parentheses
        Private field1_ As Integer
        Private field2_ As Integer

        Public ReadOnly Property Field2 As Integer
            Get
                Return (((Me.field1_))) ' Noncompliant
'                            ^^^^^^^
            End Get
        End Property
    End Class

    Class NonCompliant_OuterClass
        Private fielda_ As String
        Private fieldb_ As String

        Structure NonCompliant_NestedClass
            Private fielda_ As Integer
            Private fieldb_ As Integer

            Public Property FieldA As Integer
                Get
                    Return Me.fieldb_  ' Noncompliant
                End Get
                Set(ByVal value As Integer)
                    Me.fieldb_ = value ' Noncompliant
                End Set
            End Property
        End Structure
    End Class

    Structure Compliant_Indexer
        ' Declare an array to store the data elements.
        Private Shared ReadOnly arr As Integer() = New Integer(99) {}

        ' Define the indexer to allow client code to use [] notation.
        Default Public Property Item(ByVal i As Integer) As Integer
            Get
                Return arr(i) ' Compliant - we don't know which field to check against
            End Get
            Set(ByVal value As Integer)
                arr(i) = value
            End Set
        End Property
    End Structure

    Class CompliantClass
        Private xxx_ As Integer

        Public Property XXX As Integer
            Get
                Return xxx_
            End Get
            Set(ByVal value As Integer)
                xxx_ = value
            End Set
        End Property

        Public Property UUU As Integer
            Get
                Return xxx_ ' Compliant - no matching field name
            End Get
            Set(ByVal value As Integer)
                xxx_ = value
            End Set
        End Property

        Private _a_b_c As String
        Private abc As String
        Private yyy As String

        Public Property Abc2 As String
            Get
                Return yyy ' Compliant - multiple possible matching field names, so don't raise
            End Get
            Set(ByVal value As String)
                yyy = value
            End Set
        End Property
    End Class

    Class Compliant_ImplicitProperties
        Private firstName As String
        Private secondName As String
        Public Property FirstName2 As String = "Jane"
        Public Property SecondName2 As String
    End Class

    Class WrappedClass
        Friend field1 As Integer
        Friend field2 As Integer
    End Class

    Class Compliant_WrappedObject
        Private wrapped As WrappedClass

        Public Property Field2 As Integer
            Get
                Return wrapped.field1
            End Get
            Set
                wrapped.field1 = value
            End Set
        End Property
    End Class

    Class BaseClass
        Protected field1 As Integer
    End Class

    Class ChildClass
        Inherits BaseClass

        Private field2 As Integer

        Public Property Field1 As Integer
            Get
                Return field2 ' Compliant - aren't checking inherited fields
            End Get
            Set(ByVal value As Integer)
                field2 = value
            End Set
        End Property
    End Class

    Class MultipleOperations
        Private _foo As Integer
        Private _bar As Integer

        Public Property Foo As Integer
            Get
                Return _foo
            End Get
            Set(ByVal value As Integer)
                _bar = 1
                _foo = value ' Compliant
                _bar = 0
            End Set
        End Property
    End Class

    ' this usage is specific to MVVM Light framework
    Public Class FooViewModel
        Inherits ViewModelBase

        Private _foo As Integer
        Private _bar As Integer

        Public Property Foo As Integer
            Get
                Return Me._foo
            End Get
            Set(ByVal value As Integer)
                If Me.[Set](Me._foo, 1) Then ' Compliant, it is assigned in the Set method
                    Me._bar = 1
                End If
            End Set
        End Property
    End Class

    Public Class FooViewModelWithoutSet
        Inherits ViewModelBase

        Private _foo As Integer
        Private _bar As Integer

        Public Function MySet(ByVal x As Integer, ByVal y As Integer) As Boolean
            Return True
        End Function

        Public Property Foo As Integer
            Get
                Return Me._foo
            End Get
            Set(ByVal value As Integer)
                If MySet(Me._foo, 1) Then
                    Me._bar = 1 ' Noncompliant
                End If
            End Set
        End Property
    End Class

    Public Class MultipleStatements
        Private _foo As String
        Private _bar As String
        
        Public Property Foo As String
            Get
                If true Then
                    Throw New System.InvalidOperationException("")
                End If

                _foo = "stuff"
                Return _bar ' Noncompliant {{Refactor this getter so that it actually refers to the field '_foo'.}}
'                      ^^^^
            End Get
            Set
                If _foo.Equals(_foo) Then
                    _bar = value ' Noncompliant {{Refactor this setter so that it actually refers to the field '_foo'.}}
'                   ^^^^
                End If

            End Set
        End Property

        Public Property Bar As String
            Get
                If true Then
                    Throw New System.InvalidOperationException("")
                End If

                Me._bar = "stuff"
                Return Me._foo ' Noncompliant {{Refactor this getter so that it actually refers to the field '_bar'.}}
'                         ^^^^
            End Get
            Set
                If Me._bar.Equals(Me._foo) Then
                    Me._foo = value ' Noncompliant {{Refactor this setter so that it actually refers to the field '_bar'.}}
'                      ^^^^
                End If

            End Set
        End Property
    End Class

    Public Class SpecialUsages
        Private _foo As String

        Public ReadOnly Property Foo As String
            Get
                Dim variable = Me._foo ' Compliant, field is read
                If true Then
                    variable = (variable + variable)
                End If
                Return variable
            End Get
        End Property

        Public WriteOnly Property DoNotSet As Integer
            Set
                Throw New System.InvalidOperationException("") ' Compliant
            End Set
        End Property

        Private _tux As String

        Public ReadOnly Property Tux As String
            Get
                Return (_tux + "salt") ' Compliant
            End Get
        End Property

        Private _mux As String

        Public ReadOnly Property Mux As String
            Get
                Return _mux.Replace("x", "y") ' Compliant
            End Get
        End Property
    End Class

End Namespace

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PropertiesAccessCorrectField.cs
using System;
using GalaSoft.MvvmLight;

namespace Tests.Diagnostics
{
    class NonCompliantClass_FromRspec
    {
        private int x;
        private int y;

        public int X
        {
            get { return x; }
            set { x = value; }
        }

        public int Y
        {
            get { return x; }  // Noncompliant {{Refactor this getter so that it actually refers to the field 'y'.}}
//                       ^
            set { x = value; } // Noncompliant {{Refactor this setter so that it actually refers to the field 'y'.}}
//                ^
        }
    }

    struct NonCompliantStruct_FromRspec
    {
        private int x;
        private int y;

        public int X
        {
            get { return x; }
            set { x = value; }
        }

        public int Y
        {
            get { return x; }  // Noncompliant: field 'y' is not used in the return value
            set { x = value; } // Noncompliant: field 'y' is not updated
        }
    }


    class NonCompliant_UnderscoresInNamesAndCasing
    {
        private int yyy;
        private int __x__X; // test that underscores and casing in names are ignored
        public int XX
        {
            get { return yyy; }   // Noncompliant {{Refactor this getter so that it actually refers to the field '__x__X'.}}
//                       ^^^
            set { yyy = value; }  // Noncompliant {{Refactor this setter so that it actually refers to the field '__x__X'.}}
//                ^^^
        }

        public int _Y___Y_Y_
        {
            get { return __x__X; } // Noncompliant
//                       ^^^^^^
        }
    }

    class NonCompliant_FieldTypeIsIgnored
    {
        private int aaa;
        private string aString;

        public string AAA
        {
            get { return aString; } // Noncompliant - field called 'aaa' exists, even though type is different
//                       ^^^^^^^
            set { aString = value; } // Noncompliant
//                ^^^^^^^
        }
    }

    class NonCompliant_AssigningToExpression
    {
        private int aaa;
        private string aString;

        public string AAA
        {
            set { aString = "foo" + value; } // Noncompliant
//                ^^^^^^^
        }
    }

    partial class NonCompliant_PartialClass
    {
        private object myProperty;
    }
    partial class NonCompliant_PartialClass
    {
        private object anotherObject;
    }
    partial class NonCompliant_PartialClass
    {
        public object MyProperty
        {
            get { return this.anotherObject; } // Noncompliant
            set { this.anotherObject = value; } // Noncompliant
        }
    }


    class NonCompliant_ComplexProperty
    {
        private int field1;
        private int field2;
        private bool initialized;
        private bool isDisposed;

        public int Field1
        {
            get // Noncompliant {{Refactor this getter so that it actually refers to the field 'field1'.}}
//          ^^^
            {
                if (!this.initialized)
                {
                    throw new InvalidOperationException();
                }
                if (this.isDisposed)
                {
                    throw new ObjectDisposedException("object name");
                }

                return this.field2;
            }

            set
            {
                if (value < 0)
                {
                    throw new ArgumentOutOfRangeException(nameof(value));
                }
                this.field2 = value; // Noncompliant
//                   ^^^^^^
            }
        }
    }

    class NonCompliant_Parentheses
    {
        private int field1;
        private int field2;

        public int Field2
        {
            get { return (((this.field1))); } // Noncompliant
//                               ^^^^^^
            set
            {
                (((field1))) = value; // Noncompliant
//                 ^^^^^^
            }
        }
    }


    class NonCompliant_OuterClass
    {
        private string fielda;
        private string fieldb;

        struct NonCompliant_NestedClass
        {
            int fielda;
            int fieldb;

            public int FieldA
            {
                get { return this.fieldb; }     // Noncompliant
                set { this.fieldb = value; }   // Noncompliant
            }
        }
    }


    class Compliant_Indexer
    {
        // Declare an array to store the data elements.
        private readonly int[] arr = new int[100];

        // Define the indexer to allow client code to use [] notation.
        public int this[int i]
        {
            get { return arr[i]; }  // Compliant - we don't know which field to check against
            set { arr[i] = value; }
        }
    }

    class CompliantClass
    {
        private int xxx;

        public int XXX
        {
            get { return xxx; }
            set { xxx = value; }
        }

        public int UUU
        {
            get { return xxx; }     // Compliant - no matching field name
            set { xxx = value; }
        }

        private string _a_b_c;
        private string abc;
        private string yyy;

        public string Abc
        {
            get { return yyy; }     // Compliant - multiple possible matching field names, so don't raise
            set { yyy = value; }
        }
    }

    class Compliant_ImplicitProperties
    {
        private string firstName;
        private string secondName;

        public string FirstName { get; set; } = "Jane";
        public string SecondName { get; set; }
    }


    class WrappedClass
    {
        internal int field1;
        internal int field2;
    }
    class Compliant_WrappedObject
    {
        private WrappedClass wrapped;

        public int Field2
        {
            get { return wrapped.field1; }
            set { wrapped.field1 = value; }
        }
    }

    class BaseClass
    {
        protected int field1;
    }

    class ChildClass : BaseClass
    {
        private int field2;

        public int Field1
        {
            get { return field2; }      // Compliant - aren't checking inherited fields
            set { field2 = value; }     // Compliant
        }
    }

    class MultipleOperations
    {
        private int _foo;
        private int _bar;
        public int Foo
        {
            get
            {
                return _foo;
            }
            set
            {
                _bar = 1;
                _foo = value; // Compliant
                _bar = 0;
            }
        }
    }

    struct MultipleOperationsStruct
    {
        private int _foo;
        private int _bar;
        public int Foo
        {
            get
            {
                return _foo;
            }
            set
            {
                _bar = 1;
                _foo = value; // Compliant
                _bar = 0;
            }
        }
    }

    // this usage is specific to MVVM Light framework
    public class FooViewModel : ViewModelBase
    {
        private int _foo;
        private int _bar;

        public int Foo
        {
            get => this._foo;
            set
            {
                if (this.Set(ref this._foo, 1)) // Compliant - the Set method does the assignment
                {
                    this._bar = 1;
                }
            }
        }
    }

    public class FooViewModelWithoutSet : ViewModelBase
    {
        private int _foo;
        private int _bar;

        public bool MySet(int x, int y) => true;

        public int Foo
        {
            get => this._foo;
            set
            {
                if (MySet(this._foo, 1))
                {
                    this._bar = 1; // Noncompliant
                }
            }
        }
    }

    public class NoFieldUsage
    {
        private int foo1;
        public int Foo1 { get; }

        private int foo2;
        public int Foo2 { get; set; }

        private int foo3;
        public int Foo3
        {
            get
            {
                Bar();
                return foo1; // Noncompliant
            }
            set // Noncompliant
//          ^^^
            {
                Bar();
            }
        }

        private int foo4;
        public int Foo4 => 4;

        private int foo5;
        public int Foo5
        {
            get // Noncompliant
//          ^^^
            {
                Bar();
                return 1;
            }
        }

        void Bar() { }
    }

    public class UpdateWithOut
    {
        private int foo;
        public int Foo
        {
            set
            {
                Assign(value, out foo);
            }
        }
        private void Assign(int value, out int result)
        {
            result = value;
        }
    }

    public class UpdateWithOut2
    {
        private int foo;
        private int bar;
        public int Foo
        {
            set
            {
                Assign(value, out bar); // Noncompliant
            }
        }
        private void Assign(int value, out int result)
        {
            result = value;
        }
    }

    public class MultipleStatements
    {
        private string foo;
        private string bar;
        public string Foo
        {
            get
            {
                if (true)
                {
                    throw new System.InvalidOperationException("");
                }
                foo = "stuff";
                return bar; // Noncompliant {{Refactor this getter so that it actually refers to the field 'foo'.}}
//                     ^^^
            }
            set
            {
                if (foo.Equals(foo))
                {
                    bar = value; // Noncompliant {{Refactor this setter so that it actually refers to the field 'foo'.}}
//                  ^^^
                }
            }
        }
        public string Bar
        {
            get
            {
                if (true)
                {
                    throw new System.InvalidOperationException("");
                }
                this.bar = "stuff";
                return this.foo; // Noncompliant {{Refactor this getter so that it actually refers to the field 'bar'.}}
//                          ^^^
            }
            set
            {
                if (this.bar.Equals(foo))
                {
                    this.foo = value; // Noncompliant {{Refactor this setter so that it actually refers to the field 'bar'.}}
//                       ^^^
                }
            }
        }

    }

    public class SpecialUsages
    {
        private string _foo;

        public string Foo
        {
            get
            {
                var foo = _foo; // Compliant, field is read
                if (true)
                {
                    foo += foo;
                }
                return foo;
            }
        }

        private object baz;
        public object Baz
        {
            get => baz ?? throw new System.InvalidOperationException(""); // Compliant
            set => baz = (baz == null) ? value : throw new System.InvalidOperationException(""); // Compliant
        }

        private int doNotSet;
        public int DoNotSet
        {
            set
            {
                throw new System.InvalidOperationException(""); // Compliant, if it throws do not raise
            }
        }

        private string tux;
        public string Tux
        {
            get
            {
                return tux + "salt"; // Compliant
            }
        }

        private string mux;
        public string Mux
        {
            get
            {
                return mux.Replace('x', 'y');
            }
        }

        private string MultipleFields
        {
            get => _foo == baz ? tux : mux;
        }

        private object baz2;
        public object Baz2
        {
            get /* 123 */ => /* 456 */ throw new System.InvalidOperationException(""); // Compliant
            set => throw new System.InvalidOperationException(""); // Compliant
        }
    }
}
